src/System.Private.CoreLib/src/System/Threading/Timer.CoreCLR.cs

```diff
@@ -0,0 +1,139 @@
+        {
+            public AppDomainTimerSafeHandle()
+                : base(true)
+            {
+            }
+
+            protected override bool ReleaseHandle()
+            {
+                return DeleteAppDomainTimer(handle);
+            }
+        }
+
+        private AppDomainTimerSafeHandle m_appDomainTimer;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool SetTimer(uint actualDuration)
+        {
+            if (m_appDomainTimer == null || m_appDomainTimer.IsInvalid)
+            {
+                Debug.Assert(!m_isAppDomainTimerScheduled);
+                Debug.Assert(m_id >= 0 && m_id < Instances.Length && this == Instances[m_id]);
+
+                m_appDomainTimer = CreateAppDomainTimer(actualDuration, m_id);
+                if (!m_appDomainTimer.IsInvalid)
+                {
+                    m_isAppDomainTimerScheduled = true;
+                    m_currentAppDomainTimerStartTicks = TickCount;
+                    m_currentAppDomainTimerDuration = actualDuration;
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                if (ChangeAppDomainTimer(m_appDomainTimer, actualDuration))
+                {
+                    m_isAppDomainTimerScheduled = true;
+                    m_currentAppDomainTimerStartTicks = TickCount;
+                    m_currentAppDomainTimerDuration = actualDuration;
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+        }
+
+        // The VM calls this when a native timer fires.
+        internal static void AppDomainTimerCallback(int id)
+        {
+            Debug.Assert(id >= 0 && id < Instances.Length && Instances[id].m_id == id);
+            Instances[id].FireNextTimers();
+        }
+
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        private static extern AppDomainTimerSafeHandle CreateAppDomainTimer(uint dueTime, int id);
+
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        private static extern bool ChangeAppDomainTimer(AppDomainTimerSafeHandle handle, uint dueTime);
+
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        private static extern bool DeleteAppDomainTimer(IntPtr handle);
+
+        #endregion
+    }
+
+    // A timer in our TimerQueue.
+    internal sealed partial class TimerQueueTimer
+    {
+        internal void SignalNoCallbacksRunning()
+        {
+            object toSignal = m_notifyWhenNoCallbacksRunning;
+            Debug.Assert(toSignal is WaitHandle || toSignal is Task<bool>);
+
+            if (toSignal is WaitHandle wh)
+            {
+                Interop.Kernel32.SetEvent(wh.SafeWaitHandle);
```

Comment:
It will likely be possible to move this one function to shared code as well. I assume I will need to `#if PLATFORM_WINDOWS` and switch between `Interop.Kernel32.SetEvent(handle);` and `WaitSubsystem.SetEvent(handleValue);` on different platforms.